src/utils/FileBuffer.cpp

```diff
@@ -0,0 +1,285 @@
+
+#include <cstdlib>
+#include <ctime>
+#include <libftpp/expected.hpp>
+#include <libftpp/string.hpp>
+#include <utils/IBuffer.hpp>
+
+#include <cstddef>
+#include <cstdio>
+#include <fstream>
+#include <ios>
+#include <iostream>
+#include <string>
+#include <sys/stat.h>
+
+/* ************************************************************************** */
+// INIT
+
+const char* const FileBuffer::errOpen = "FbException: open failed";
+const char* const FileBuffer::errSeek = "FbException: seek failed";
+const char* const FileBuffer::errFileEmpty = "FbException: file is empty";
+const char* const FileBuffer::errFileNotOpened =
+  "FbException: file is not opened";
+const char* const FileBuffer::errEof = "FbException: tried to get end of file";
+const char* const FileBuffer::errRead = "FbException: read failed";
+const char* const FileBuffer::errWrite = "FbException: write failed";
+
+/* ************************************************************************** */
+// PUBLIC
+
+FileBuffer::FileBuffer()
+  : _size(0)
+{
+}
+
+FileBuffer::~FileBuffer()
+{
+  if (_fs.is_open()) {
+    _fs.close();
+    (void)std::remove(_fileName.c_str());
+  }
+}
+
+IBuffer::ExpectChr FileBuffer::get()
+{
+  if (!_fs.is_open()) {
+    return ft::unexpected<BufferException>(errFileNotOpened);
+  }
+
+  const int chr = _fs.get();
+  if (_fs.fail()) {
+    if (_fs.eof()) {
+      return ft::unexpected<BufferException>(errEof);
+    }
+    return ft::unexpected<BufferException>(errRead);
+  }
+  return chr;
+}
+
+IBuffer::ExpectChr FileBuffer::peek()
+{
+  if (!_fs.is_open()) {
+    return ft::unexpected<BufferException>(errFileNotOpened);
+  }
+
+  const int chr = _fs.peek();
+  if (_fs.fail()) {
+    if (_fs.eof()) {
+      return ft::unexpected<BufferException>(errEof);
+    }
+    return ft::unexpected<BufferException>(errRead);
+  }
+  return chr;
+}
+
+IBuffer::ExpectVoid FileBuffer::seek(std::size_t pos)
+{
+  if (!_fs.is_open()) {
+    return ft::unexpected<BufferException>(errFileNotOpened);
+  }
+
+  _fs.seekg(static_cast<std::streamoff>(pos));
+  if (_fs.fail()) {
+    return ft::unexpected<BufferException>(errSeek);
+  }
+  return ExpectVoid();
+}
+
+IBuffer::ExpectVoid FileBuffer::append(const std::string& data)
+{
+  if (!_fs.is_open()) {
+    return ft::unexpected<BufferException>(errFileNotOpened);
+  }
+
+  // Go to end and add new data
+  _fs.seekp(0, std::ios::end);
+  if (_fs.fail()) {
+    return ft::unexpected<BufferException>(errSeek);
+  }
+
+  _fs << data;
+  _calcSize();
+  return ExpectVoid();
+}
+
+IBuffer::ExpectVoid FileBuffer::append(const FileBuffer::Container& buffer,
+                                       long bytes)
+{
+  if (!_fs.is_open()) {
+    return ft::unexpected<BufferException>(errFileNotOpened);
+  }
+
+  // Go to end and add new data
+  _fs.seekp(0, std::ios::end);
+  if (_fs.fail()) {
+    return ft::unexpected<BufferException>(errSeek);
+  }
+
+  _fs.write(buffer.data(), bytes);
+  if (_fs.fail() || _fs.bad()) {
```

Comment:
Badbit gets checked in `fail()` anw.
```suggestion
  if (_fs.fail()) {
```